Starting with recipes

dmp42 opened the pull request:
- apache recipe, including group level fine-grain authorization (enforce -B for @stevvooe)
- streamlined nginx recipe a bit to match better with Apache
- recipes file

Validated and fully tested accompanying shell scripts.

Fixes #855 and starts the recipe effort.

cc @docker/distribution-maintainers

dmp42 commented:
cc @thaJeztah for the usual laser-doc-review

RichardScothern commented on docs/apache.md:
```diff
@@ -0,0 +1,228 @@
+
+Usually, that includes enterprise setups using LDAP/AD on the backend and a SSO mechanism fronting their internal http portal.
+
+## Alternatives
```
Make this a sub header?
### Alternatives

dmp42 replied:
Agreed.

RichardScothern commented on docs/apache.md:
```diff
@@ -0,0 +1,228 @@
+
+We also implement push restriction (to a limited user group) for the sake of the exemple. Again, you should modify this to fit your mileage. 
+
+## Gotchas
```
Also a sub header?

dmp42 replied:
Agreed.

RichardScothern commented on docs/apache.md:
```diff
@@ -0,0 +1,228 @@
+
+## Setting things up
+
+Read again the requirements.
```
Which requirements?  Could you link to them?

dmp42 replied:
Fixed. Also in nginx.md

RichardScothern commented on docs/apache.md:
```diff
@@ -0,0 +1,228 @@
+
+ServerAdmin you@example.com
+
+ErrorLog /proc/self/fd/2
```
Maybe note this is Linux/Unix only?

dmp42 replied:
Mmmm... I see your point.
Now, this is really meant to be used with a specific hub official image, so, I think we are ok with that for now (people running this kind of s on windows is still a long shot).

RichardScothern replied:
Yes, probably not a big deal, although BSD doesn't have /proc either

stevvooe replied:
Maybe, something like this: `ErrorLog | sh -c 'cat 1>&2'`.

dmp42 replied:
I can't figure out a way to tame Apache to do that (re: http://httpd.apache.org/docs/2.4/mod/core.html#errorlog )

stevvooe replied:
This example doesn't work:

```
ErrorLog "|/usr/local/bin/httpd_errors"
```

This:

```
ErrorLog "|sh -c 'cat 1>&2'"
```

dmp42 replied:
Neither work for me...

thaJeztah replied:
I wonder if you should completely replace the default configuration file, or just add this file to the "extras" dir and include it in the main configuration file.

This is what the standard configuration file uses for logging;

https://github.com/docker-library/httpd/blob/502ecf54a9db85f556028036973a9bce72eae8e8/2.4/Dockerfile#L57-L58

dmp42 replied:
The default config file is no good IMO - eg: not specific enough, and not secure enough. But yeah, basically they are doing the same thing as far as logging is concerned.

RichardScothern commented on docs/apache.md:
```diff
@@ -0,0 +1,228 @@
+  SSLCipherSuite RC4-SHA:HIGH
+  SSLHonorCipherOrder on
+
+  Header always set "Docker-Distribution-Api-Version" "registry/2.0"
```
Comment for this section?

RichardScothern commented on docs/apache.md:
```diff
@@ -0,0 +1,228 @@
+  SSLProtocol all -SSLv2 -SSLv3 -TLSv1
+  SSLCipherSuite RC4-SHA:HIGH
+  SSLHonorCipherOrder on
+
```
Comment for this section

dmp42 replied:
Calling @diogomonica and @NathanMcCauley to the rescue on this: can you guys validate my tls babbling here is remotely proper?

aaronlehmann replied:
You definitely shouldn't use RC4.

RichardScothern commented on docs/apache.md:
```diff
@@ -0,0 +1,228 @@
+
+Now, start your stack:
+
+    docker-compose up -d`
```
extraneous (or missing) `

dmp42 replied:
Fixed.

RichardScothern commented:
A few nits, but a tasty recipe.

dmp42 commented:
@RichardScothern PTAL

Also, do you think I should block PATCH as well? Do you know why it's not failing currently?

dmp42 commented:
@aaronlehmann PTAL on the SSL gymnastics.

aaronlehmann commented:
The SSL stuff looks good to me. I was going to recommend the same page you cited.

RichardScothern commented:
LGTM

dmp42 commented:
Updated nginx TLS settings as well.
Will do some TLS automated testing in a few to validate all that.

Added PATCH to the list of methods in the Apache config.
Couldn't figure out a way to have better Apache Log statements but then...

That should be good enough for a start at this. Going to chase Apache struggling users around and have them wagtest it.
